homeassistant/components/weather/__init__.py

```diff
@@ -115,36 +190,84 @@ class WeatherEntity(Entity):
     _attr_visibility: float | None = None
-    _attr_visibility_unit: str | None = None
-    _attr_precipitation_unit: str | None = None
+    _attr_visibility_unit: None = (
+        None  # Subclasses of WeatherEntity should not set this
+    )
+    _attr_native_visibility_unit: str | None = None
+    _attr_precipitation_unit: None = (
+        None  # Subclasses of WeatherEntity should not set this
+    )
+    _attr_native_precipitation_unit: str | None = None
     _attr_wind_bearing: float | str | None = None
     _attr_wind_speed: float | None = None
-    _attr_wind_speed_unit: str | None = None
+    _attr_wind_speed_unit: None = (
+        None  # Subclasses of WeatherEntity should not set this
+    )
+    _attr_native_wind_speed_unit: str | None = None
+
+    _weather_option_temperature_uom: str | None = None
+    _weather_option_pressure_uom: str | None = None
+    _weather_option_visibility_uom: str | None = None
+    _weather_option_precipitation_uom: str | None = None
+    _weather_option_wind_speed_uom: str | None = None
+
+    async def async_internal_added_to_hass(self) -> None:
+        """Call when the sensor entity is added to hass."""
+        await super().async_internal_added_to_hass()
+        if not self.registry_entry:
+            return
+        self.async_registry_entry_updated()
 
     @property
     def temperature(self) -> float | None:
         """Return the platform temperature in native units (i.e. not converted)."""
         return self._attr_temperature
 
     @property
-    def temperature_unit(self) -> str:
+    def native_temperature_unit(self) -> str | None:
         """Return the native unit of measurement for temperature."""
-        return self._attr_temperature_unit
+        if hasattr(self, "_attr_native_temperature_unit"):
+            return self._attr_native_temperature_unit
+        return None
+
+    @property
+    def temperature_unit(self) -> str:
+        """Return the unit of measurement for temperature."""
+        if self._weather_option_temperature_uom:
+            return self._weather_option_temperature_uom
+
+        return self.hass.config.units.temperature_unit
 
     @property
     def pressure(self) -> float | None:
         """Return the pressure in native units."""
         return self._attr_pressure
 
     @property
-    def pressure_unit(self) -> str | None:
+    def native_pressure_unit(self) -> str | None:
         """Return the native unit of measurement for pressure."""
-        return self._attr_pressure_unit
+        if hasattr(self, "_attr_native_pressure_unit"):
+            return self._attr_native_pressure_unit
+        return None
+
+    @property
+    def pressure_unit(self) -> str | None:
+        """Return the unit of measurement for pressure."""
+        if self._weather_option_pressure_uom:
+            return self._weather_option_pressure_uom
+
+        if self.hass.config.units.is_metric:
+            return PRESSURE_HPA
+        return PRESSURE_INHG
```

Comment:
I think this would be a breaking change since pressure is not currently converted unless entities opt-in by setting `pressure_unit`?

To make the code clearer, I'd suggest to return `None` if `self.native_pressure_unit` is `None`